src/StakingNode.sol

```diff
@@ -399,56 +463,90 @@ contract StakingNode is IStakingNode, StakingNodeEvents, ReentrancyGuardUpgradea
+            if (withdrawals[i].scaledShares.length != 1 || withdrawals[i].strategies.length != 1 || withdrawals[i].strategies[0] != beaconChainETHStrategy) {
                 revert InvalidWithdrawal();
             }
-            tokens[i] = new IERC20[](1);
+            tokens[i] = new IERC20V4[](1);
             receiveAsTokens[i] = false;
-            totalWithdrawalAmount += withdrawals[i].shares[0];
+            
+            totalWithdrawableShares += _decreaseQueuedSharesOnCompleteQueuedWithdrawalAndGetTotalWithdrawableShare(delegationManager, withdrawals[i]);
         }
 
-        IDelegationManager delegationManager = IDelegationManager(address(stakingNodesManager.delegationManager()));
-
         // Complete withdrawals with receiveAsTokens = false
-        delegationManager.completeQueuedWithdrawals(withdrawals, tokens, middlewareTimesIndexes, receiveAsTokens);
+        delegationManager.completeQueuedWithdrawals(
+            withdrawals, 
+            tokens, 
+            receiveAsTokens
+        );
 
-        // Decrease queued shares amount
-        queuedSharesAmount -= totalWithdrawalAmount;
+        emit CompletedQueuedWithdrawals(withdrawals, totalWithdrawableShares, 0);
+    }
 
-        emit CompletedQueuedWithdrawals(withdrawals, totalWithdrawalAmount, 0);
+    /**
+     * @notice Decreases the queued shares on complete queued withdrawals and returns the total withdrawable shares
+     * @param delegationManager The delegation manager
+     * @param withdrawal The withdrawal struct
+     * @return totalWithdrawableShareForWithdrawalRoot The total withdrawable shares for the withdrawal root
+     */
+    function _decreaseQueuedSharesOnCompleteQueuedWithdrawalAndGetTotalWithdrawableShare(
+        IDelegationManager delegationManager,
+        IDelegationManager.Withdrawal calldata withdrawal
+    ) internal returns (uint256) {
+        uint256 totalWithdrawableShareForWithdrawalRoot = 0;
+        bytes32 withdrawalRoot = delegationManager.calculateWithdrawalRoot(withdrawal);
+        uint256 withdrawableSharesForWithdrawalRootValue = withdrawableSharesForWithdrawalRoot[withdrawalRoot];
+
+            if (withdrawableSharesForWithdrawalRootValue != 0) {
+                // If the withdrawableSharesForWithdrawalRootValue are not 0, that means withdrawal root exists in withdrawableSharesForWithdrawalRoot
+                // which means it was queued after the upgrade
+                // so we need to subtract the shares from queuedSharesAmount and set the withdrawableSharesForWithdrawalRoot to 0
+                totalWithdrawableShareForWithdrawalRoot = withdrawableSharesForWithdrawalRootValue;
+                withdrawableSharesForWithdrawalRoot[withdrawalRoot] = 0;
+                queuedSharesAmount -= withdrawableSharesForWithdrawalRootValue;
+            } else {
+                // If the withdrawableSharesForWithdrawalRootValue are 0, that means withdrawal root doesn't exist in withdrawableSharesForWithdrawalRoot
```

Comment:
does this function assume that the function that calls it reverts if the withdrawal has already been completed or does not exist? 

withdrawableSharesForWithdrawalRootValue is 0 if it was never set or is set here to -